### PR TITLE
Add ERC: Composable Security Middleware Hooks

### PR DESCRIPTION
**Pull request by peersky**

### CODE REVIEW

File: ERCS/erc-0000.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+---
+eip: 0000
```
Comment (abcoathup) [inline ERCS/erc-0000.md:2]:
```suggestion
eip: 7746
```
Assigning next sequential EIP/ERC/RIP number.

Please also update the filename.

File: ERCS/erc-0000.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+title: Security Layers
```
Comment (abcoathup) [inline ERCS/erc-0000.md:3]: Is there a more descriptive title you could use?  Composable runtime security checks

File: ERCS/erc-0000.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+description: An interface for composable, runtime security checks in smart contracts.
+author: Tim Pechersky (@peersky) <t@peersky.xyz>
+discussions-to: https://ethereum-magicians.org/t/layered-smart-contract-security/19471
```
Comment (abcoathup) [inline ERCS/erc-0000.md:6]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7746-layered-smart-contract-security/19471
```

File: ERCS/erc-0000.md (hunk 2 above)
Comment (peersky) [reply ERCS/erc-0000.md]: Changed it to `Composable Security Middleware Hooks` , hope this sounds more self-explanatory

File: assets/erc-7746/.gitignore
Comment (SamWilsn) [inline assets/erc-7746/.gitignore:1]: This is a pretty big ignore file for a reference implementation :P

File: assets/erc-7746/hardhat.config.ts
Comment (SamWilsn) [inline assets/erc-7746/hardhat.config.ts:1]: Please remove build system files.

File: ERCS/erc-7746.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+**Strict Validation**: Reverts on validation failure ensure a fail-safe mechanism, preventing execution of potentially harmful transactions.
+
+**Gas Costs**: Layers naturally will have gas costs associated with their execution. However, the benefits of enhanced security and modularity outweigh these costs, especially as blockchain technology continues to evolve and we expect gas costs to decrease over time.
+
+## Reference Implementation
+
+A reference implementation of the ILayer interface and a sample protected contract can be found in the repository:
+[../assets/erc-7746/](../assets/erc-7746/)
+
```
Comment (SamWilsn) [inline ERCS/erc-7746.md:90]: I'd like to see a bit of a walkthrough of the reference implementation. For example, which files are important to the standard, and which are just there for testing.

**Comment by peersky** [comment]
@SamWilsn thanks for previous input. Could you please review new changes, when you have time?  I've addressed all of items since last time we spoke

File: ERCS/erc-7746.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        address sender,
+        uint256 value,
+        bytes memory data,
+        bytes memory beforeCallResult
+    ) external;
+}
+
+```
+
+A protected contract MAY integrate security layers by calling the beforeCallValidation function before executing its logic and the afterCallValidation function afterwards. Multiple layers can be registered and executed in a defined order. The protected contract MUST revert if any layer reverts.
```
Comment (SamWilsn) [inline ERCS/erc-7746.md:74]:
```suggestion
A protected contract MAY integrate security layers by calling the `beforeCallValidation` function before executing its logic and the `afterCallValidation` function afterwards. Multiple layers can be registered and executed in a defined order. The protected contract MUST revert if any layer reverts.
```

File: ERCS/erc-7746.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+In the [../assets/eip-7746/ILayer.sol](../assets/eip-7746/ILayer.sol) a reference interface is provided.
+
+In this test, a [Protected.sol](../assets/eip-7746/test/Protected.sol) contract is protected by a [RateLimitLayer.sol](../assets/eip-7746/test/RateLimitLayer.sol) layer. The RateLimitLayer implements the ILayer interface and enforces a rate which client has configured.
+The Drainer simulates a vulnerable contract that acts in a malicious way. In the `test.ts` The Drainer contract is trying to drain the funds from the Protected contract. It is assumed that Protected contract has bug that allows partial unauthorized access to the state.
+The RateLimitLayer is configured to allow only 10 transactions per block from same sender. The test checks that the Drainer contract is not able to drain the funds from the Protected contract.
+
+github repository with example can be seen here
+```
+https://github.com/peersky/smart-contract-layers/
+```
```
Comment (SamWilsn) [inline ERCS/erc-7746.md:98]: Please remove this external link.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**